dev/camdenhells/REQUIREMENTS.md

```diff
@@ -324,41 +350,75 @@ This does not preclude performance impacts; clearly there will be limits, but ha
+  
+  * Automated unit/integration test harness and tests 
+  * For all public API and internal interface methods except those:
+    * With significant historical runtime event dependencies or behavioral states, and/or
+    * Whose success cannot be (at least partially) reasonably determined as value assertions
+  * Includes e.g. contructors, simple get/set accessors, deterministic functions/transforms
+  * Excludes e.g. frame draws, client callbacks
+  
+## 4.3. Run-time
+
+For engine and app development and integration testing, and for debugging support issues
+
+1. Provide unified logging
+  * The usual levels - verbose, debug, info, warning, error, wtf
+  * All messages should identify the class and method
+  * Core engine/rendering component (scenes, scene view, render pass, etc.) messages should also identify the instance
+  * All warning and error messages should include specific causes and data values if appropriate
+  * All error messages should be accompanied by a thrown exception
+2. Provide debug-level method tracing messages 
+  * For key lifecycle event (create/init/start/run/pause/stop/destroy etc.) handlers of engine/rendering components
+    * Includes all active components: scenes, scene views, render passes, etc.
+    * Does not include passive model objects, textures, resources 
+  * Tracing dynamically enabled per instance; disabled by default  
+3. Use logging and trace messaging judiciously to avoid excessive performance impacts or major frame delays
+  * Comment out or delete developmental log messages when no longer needed
+  * Avoid frame-rate messages
+    * Accumulate/integrate/summarize per-frame data, and log e.g. once per second or so
+4. API and internal interface usage validation, e.g. debug-only assertions/exceptions
+  * Check input parameters for all public API methods and all internal interface methods 
+  * Check system and instance state for lifecycle event handlers
+  * Check thread for non-thread-safe methods
+  * Except for error-level log messages, strip all logging, tracing, and assertion code from release builds
+    * Remove code rather than simply using conditional checks
 
 # 5. Modularity Requirements
 
-  * Identify and implement separate library modules for optional/lower-frequency features
-    * Core module vs add-ons:
-      * VR? AR?
-      * Modifiers, animations, rendering effects?
-      * Wear? TV?
+  * Identify and implement separate library modules for optional/lower-frequency features [2.1]
+    * Core module vs add-ons (list TBD):
+      * VR 
+      * AR
+      * Modifiers, animations, rendering effects
+      * Font-to-texture
+      * Gestures
+      * Wear
+      * TV?
       * Wallpaper
-  * Still provide all-in-one module?
-  * Publish debug/development version in addition to release version of each module?
+  * Always provide all-in-one module in addition to core plus add-on modules
+  * Publish debug/development version in addition to release version of each module
```

Comment:
All snapshots are debug/development versions. Only tagged versions will be built as true releases (stripping unnecessary code, etc).

Reply:
Ok, review to make sure I got your drift...